storage.engine: add version file to store directory

BramGruneir opened the pull request:
This adds a new file VERSION to all store directories. To facilitate this, I've
added a new proto to keep track the format. This has a tiny bit of future
proofing to allow us to have multiple database types with different data
versions as well as a minimum value we support.

When needed, having this file will allow us to automatically perform migrations
during the call to rocksDB.Open().

Fixes #2718.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5693)

<!-- Reviewable:end -->

BramGruneir committed:
storage.engine: add version file to store directory

This adds a new file COCKROACHDB_VERSION to all store directories. When needed,
having this file will allow us to automatically perform migrations during the
call to rocksDB.Open().

Fixes #2718.

tamird commented:
Note that this will likely have a write skew anomaly with #5694.

petermattis commented:
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/engine/version.proto, line 21 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE3Gke8lt1immEn2oVG:-KE3Gke8lt1immEn2oVH:1557918369) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc0e871949f80364c80ad30c9813449f2e51482/storage/engine/version.proto#L21)):*
`s/MVCCMetadata/Version/g`

Do you think a proto is overkill here? I was imagining a text file containing the version number so that it was human readable.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/engine/version.proto, line 21 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE3Gke8lt1immEn2oVG:-KE3HVTV9WE3Db1d7HQD:-824133543) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc0e871949f80364c80ad30c9813449f2e51482/storage/engine/version.proto#L21)):*
I can convert it to just a go struct.  I just kind of like the security of a proto.
Also, the file is human readable.  It's json.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/engine/version.proto, line 21 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE3Gke8lt1immEn2oVG:-KE3MkSjH_Hvp2aP49ZP:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc0e871949f80364c80ad30c9813449f2e51482/storage/engine/version.proto#L21)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/version.go, line 26 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6S3pAho6CcYcLzFT0:-KE6S3pBpZpVDbnrU_qG:217560246) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L26)):*
s/VERSION/COCKROACHDB_VERSION/ to guard against e.g. rocksdb introducing its own version file. 

---

*[storage/engine/version.go, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6Tid1mJ696GW48WLK:-KE6Tid1mJ696GW48WLL:-838387156) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L29)):*
`versionRocksDBMinimum` should be set to `versionNoFile`, so you don't need to special case `versionNoFile` when loading. Then writing the first version file can simply be the migration from `versionNoFile` to `versionRocksDBCurrent`. 

---

*[storage/engine/version.go, line 35 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6UN8PSBjGyA0y6r2R:-KE6UN8PSBjGyA0y6r2S:-1078689353) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L35)):*
In the event that we move away from rocksdb, the version file will parse successfully as if it had a rocksdb version of zero. This seems like an invitation for special-cases and bugs, so I'd rather have one top-level version field that we'll be able to continue using no matter what changes we make. We can always introduce more detailed version information when/if we need it.

---

*[storage/engine/version.go, line 43 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6SdnoKH4QhVRyVUqd:-KE6SdnoKH4QhVRyVUqe:720034109) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L43)):*
Comment doesn't match code. I think it's fine to make `versionNoFile` zero (except for the issue in the previous comment) and then you can use `Version{}` instead of `noVersionFile`. 

---

*[storage/engine/version.go, line 51 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6SHdZKob_cIQlPRog:-KE6SHd_fpOTaRTiWaIQ:-2089695684) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L51)):*
Use `filepath.Join` instead of manually constructing the path.

---

*[storage/engine/version.go, line 57 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6SQJdZwZsS8TdDRTM:-KE6SQJdZwZsS8TdDRTN:2013149637) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L57)):*
Why is this function name plural?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/version.go, line 26 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6S3pAho6CcYcLzFT0:-KE716PmqS2YykTg7ugc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L26)):*
Done.

---

*[storage/engine/version.go, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6Tid1mJ696GW48WLK:-KE73h5Vt2jXe_OX2BB-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L29)):*
Done.

---

*[storage/engine/version.go, line 35 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6UN8PSBjGyA0y6r2R:-KE71SZPrW2foKnwk2ot:1357023527) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L35)):*
Done.  

---

*[storage/engine/version.go, line 43 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6SdnoKH4QhVRyVUqd:-KE73qqboRiCJLWSZ1nN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L43)):*
Done.

---

*[storage/engine/version.go, line 51 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6SHdZKob_cIQlPRog:-KE77DkBXOMXPIQFgPVo:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L51)):*
Done.

---

*[storage/engine/version.go, line 57 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6SQJdZwZsS8TdDRTM:-KE77X-Gfb4AujCht1T4:-497468049) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L57)):*
Fixed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7IMhD8M4fEe1l3EsW:1554177296) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
why the `ver < versionCurrent` check? just to avoid the no-op case? i'd remove this.

---

*[storage/engine/rocksdb_test.go, line 175 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7I_wNErq_mErtjmPy:-KE7I_wNErq_mErtjmPz:1247898151) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb_test.go#L175)):*
the what?

---

*[storage/engine/rocksdb_test.go, line 200 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7ImnDiskFaE1a-Dh5:-KE7ImnDiskFaE1a-Dh6:-2134742054) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb_test.go#L200)):*
i know it's copied, but the 512<<20 and minmemtablebudget could both be zero, this is just misdirection currently

---

*[storage/engine/rocksdb_test.go, line 202 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7Ic0aHCZt1f7SyDWO:-KE7Ic0aHCZt1f7SyDWP:2043517111) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb_test.go#L202)):*
why is this deferred?

---

*[storage/engine/version.go, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6Tid1mJ696GW48WLK:-KE7GEsi7XYmGBj0J2z2:850816011) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L29)):*
why not set it to `versionNoFile` rather than `0`?

---

*[storage/engine/version.go, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GkaqhSLigBbQeN9R:-KE7GkaqhSLigBbQeN9S:-784003051) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L34)):*
what's the intention with using json here? if it's for the future flexibility of adding more fields, we should use a proto? if not, just strconv.FormatInt?

---

*[storage/engine/version.go, line 35 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GNOP_hay9dy4Zbif:-KE7GNOP_hay9dy4Zbig:-999346021) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L35)):*
pretty sure this json struct tag does nothing

---

*[storage/engine/version.go, line 44 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GWPKBNGDTziL3E8K:-KE7GWPKBNGDTziL3E8L:-307763257) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L44)):*
comment wrong

---

*[storage/engine/version.go, line 52 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7G_g2msRV3vT1FHMX:-KE7G_g2msRV3vT1FHMY:1772106906) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L52)):*
s/0/versionNoFile/

---

*[storage/engine/version.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GzrPdUv0hCwHCWpT:-KE7GzrPdUv0hCwHCWpU:-1747155328) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L66)):*
how can this happen?

---

*[storage/engine/version.go, line 69 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7HHnco9tVzGfuqQzn:-KE7HHnco9tVzGfuqQzo:-1738930188) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L69)):*
not necessary - `ioutil.WriteFile`: WriteFile writes data to a file named by filename. If the file does not exist, WriteFile creates it with permissions perm; otherwise WriteFile truncates it before writing.

---

*[storage/engine/version_test.go, line 70 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7HdeADRQtcqiKWgHc:-KE7HdeADRQtcqiKWgHd:735089416) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version_test.go#L70)):*
prefer `:=`

---

*[storage/engine/version_test.go, line 71 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7Hc11nD7yyVDvSnkf:-KE7Hc12eT-FD1nMOlkX:-435464181) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version_test.go#L71)):*
weird line break

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 4 files reviewed at latest revision, 19 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7W9SnR-4UZtkScAFo:963691845) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
If we don't have a file yet, we should write the version file info.  The last thing we want to to re-write the version file each time.

---

*[storage/engine/rocksdb_test.go, line 175 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7I_wNErq_mErtjmPy:-KE7W_BwOasArkc1GjeR:-1054325829) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb_test.go#L175)):*
removed

---

*[storage/engine/rocksdb_test.go, line 200 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7ImnDiskFaE1a-Dh5:-KE7WlBP3v1dMXMs4PQf:696732767) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb_test.go#L200)):*
If memtableBudget < minmemtablebudget, it will throw an error.  You're right about the cache size, I've change it to 0.

---

*[storage/engine/rocksdb_test.go, line 202 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7Ic0aHCZt1f7SyDWO:-KE7XAjnOKZ45sVh8G1Q:784065414) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb_test.go#L202)):*
I think I had something afterwards originally, removed the defer.

---

*[storage/engine/version.go, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE6Tid1mJ696GW48WLK:-KE7YwOnPTsE3FFECeiu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f0d48ee4e19d2e03ed170e5d322147e596827d/storage/engine/version.go#L29)):*
Done.

---

*[storage/engine/version.go, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GkaqhSLigBbQeN9R:-KE7XHt3Y4yWOd0TuHI3:-303375726) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L34)):*
Ok, I had this in a proto and Pete suggested that that was overkill.  So I removed the proto.  I don't care either way, but I do like the idea of having a format we can expand on and is human readable.  So I stand by the use of JSON, but I don't care if we go proto or not.

---

*[storage/engine/version.go, line 35 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GNOP_hay9dy4Zbif:-KE7XgEULPgA9SxNkVKE:-2025867346) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L35)):*
I'm pretty sure you're right.  Removed.

---

*[storage/engine/version.go, line 44 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GWPKBNGDTziL3E8K:-KE7Xrk_9qaGp8oXzc4H:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L44)):*
fixed

---

*[storage/engine/version.go, line 52 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7G_g2msRV3vT1FHMX:-KE7Xxi93kxZmaAg84IJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L52)):*
Done.

---

*[storage/engine/version.go, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GzrPdUv0hCwHCWpT:-KE7XzsM8Sj3Oy5GQsMf:-50347308) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L66)):*
it shouldn't, old code, removed.

---

*[storage/engine/version.go, line 69 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7HHnco9tVzGfuqQzn:-KE7YjriqT2VgO_I_qdk:-2068629922) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L69)):*
cool, removed

---

*[storage/engine/version_test.go, line 70 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7HdeADRQtcqiKWgHc:-KE7YsDajVakxRzhTvtu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version_test.go#L70)):*
Done.

---

*[storage/engine/version_test.go, line 71 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7Hc11nD7yyVDvSnkf:-KE7YsbQund7KBLYRJe6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version_test.go#L71)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7bTV-wBR4SetmuZm6:-1136459951) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
i'm confused. what's the harm in rewriting this file each time?

---

*[storage/engine/rocksdb_test.go, line 202 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7Ic0aHCZt1f7SyDWO:-KE7bmKR5BI6HoHy81oT:-519577756) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb_test.go#L202)):*
heh you could
```
defer rocksdb.Close()
return rocksdb.Open()
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7g8xGThEwWXUZ6rDu:1540361565) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
The harm is minimal, but there's no reason to do so.  Why do you want to write a new file each time when the current one is sufficient?

---

*[storage/engine/rocksdb_test.go, line 202 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7Ic0aHCZt1f7SyDWO:-KE7gT5G9OuI-nytA1oe:-183729665) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb_test.go#L202)):*
nice :)  Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7iNOhpdtR6qrHG-s5:1640507303) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
Because "ver < versionCurrent" wasn't obviously correct and it seems easier to remove it than to convince myself of its correctness since the outcome must be equivalent.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7j22v8Jb6IXYepfDv:369653400) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
How about if I change it to the explicit case in which we have no file?  ver == versionNoFile ?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7jQ_wzEl2VVs7gWs4:1145911872) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
seems that would be wrong when we implement the first migration.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7jzU7u-FbwsgsfLwl:-2077688187) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
Ok, I'm going to leave this as is, but I'm going to clean up the comment above it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/engine/rocksdb.go, line 148 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7IMhD8M4fEe1l3EsV:-KE7kgD9SpBn44xamt7p:-749190113) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/rocksdb.go#L148)):*
`ver < versionCurrent` seems clear enough to me. We should avoid rewriting the file unnecessarily.  

Also note that only the most drastic migrations should actually use this mechanism, since we can't make the write of the version file atomic with rocksdb operations. We'll have a version key inside the rocksdb instance for most migrations; this file would only be for changes that affect fundamental rocksdb operation (like a custom comparator). 

---

*[storage/engine/version.go, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7GkaqhSLigBbQeN9R:-KE7lRn04hxRHeIkJI1J:673867085) ([raw file](https://github.com/cockroachdb/cockroach/blob/df6761f0277452659a15b3432366ea5bdfa11f7a/storage/engine/version.go#L34)):*
I think it's good for this to be human-readable, so JSON is better than proto here. I think a plain number would be fine too, but the flexibility doesn't hurt.

---

*[storage/engine/version.go, line 70 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7mWaYAZKjl8Q8sgSf:-KE7mWaYAZKjl8Q8sgSg:1145351412) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9b3eb9399d1967b69716897b00452136714150f/storage/engine/version.go#L70)):*
`ioutil.WriteFile` is not atomic. Add a TODO to replace this with an atomic rename before we have our first migration, since it would be bad if the version file got lost due to an ill-timed crash.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-:-KE7ngnZJrdpajaIpmMW:676834233)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/engine/version.go, line 70 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE7mWaYAZKjl8Q8sgSf:-KE8Cn1rrOFawDuorG7-:1358627525) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9b3eb9399d1967b69716897b00452136714150f/storage/engine/version.go#L70)):*
I've updated the code so it is now atomic.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/engine/version.go, line 28 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE8PUVjMo60W26YBg6m:-KE8PUVjMo60W26YBg6n:-630923313) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f364844e86cd18d9587efbcc5a0c4622cf73f1c/storage/engine/version.go#L28)):*
shouldn't this be an enumerated type? sorry for late comment.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/engine/version.go, line 28 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693#-KE8PUVjMo60W26YBg6m:-KECI8RaI5OO3HqNuwse:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f364844e86cd18d9587efbcc5a0c4622cf73f1c/storage/engine/version.go#L28)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5693)*
<!-- Sent from Reviewable.io -->
